### ISSUE TITLE
Make each PR generate its own staging hub

### ISSUE THREAD

**Issue by yuvipanda** [berkeley-dsep-infra/datahub#660]
This repo has a lot of hubs, and the number is gonna keep growing. Since there's only one staging and prod branch, this causes issues.

The *right* solution here is that every PR generates its *own* entirely new hub, that can then be tested by users to make sure it works right. This lets people work independently, and test at a much faster pace. We can also do this with a bot so not every PR makes staging hubs for every hub we run. It also lets us do access control to see who can create a staging hub.

Things to sort out in this would be:

1. How to deal with authentication?
2. Who is authorized? I think we can just say anyone who can merge stuff into this is authorized.
3. When do staging hubs get torn down? Once a PR is merged, or after a few hours I'd say.

I think we need this if we gonna keep everything in this one repo, and I *think* we should keep everything in this one repo. The other option is to split it into multiple repos, but I think that comes with a level of conceptual overhead that's a little much - different places to file issues, look at docs, etc.

**Comment by yuvipanda** [comment]
Other purely social solutions are to have some sort of lock that must be aquired by someone before they can merge to staging. This lock is then valid for the next N hours, or until staging has been merged into prod. If staging has not been merged into prod by N hours, changes to staging will be automatically reverted.

This pretty much serializes access to staging merges, and makes it clear when something is being merged. It also technically enforces a social convention - if you merge to staging, you should merge to prod soon - or revert your merges to staging. This prevents us from blocking others who wanna merge to staging *unintentionally* (intentional blocking is ok), or taking other people's changes to prod.

I actually like this method much more than the more technical 'multiple staging location' method. We probably don't have enough traffic in deployements to really benefit from multiple staging hubs, and this kinda social 'lock' situation prevents accidents.

**Comment by ryanlovett** [comment]
I like the multiple staging hub idea best. How would dns work though?

One issue with the social solution that may challenge this setup: someone ("the requester") asks a merger to make a change, but the merger doesn't know how to test this change. For example a DS staffer will ask me to add a library I've never used. Currently I will ask the the requester to test and they will do so, but not within N hours. Either

1. this is okay in the new setup. there is a timeout, and the requester must make a new request and the merger must make a new merge, or
2. the merger interacts with the requester and they determine a suitable time window to merge/test, or
3. we empower and encourage the requester to be the merger.

**Comment by felder** [comment]
I don't think the social solution will work for reasons ryan has indicated as well as it still limits the number of people who can test simultaneously.  Most of the time that may not be a big deal, but as more courses come online there will be increasing amounts of contention.  So I think the multiple staging hub idea, while more complex, solves the problem of concurrency for good.

Regarding your questions:

1.  Can you tell me what the authentication challenge is here and how it differs from the current solution?

2.  Makes sense to me.

3.  When the PR is merged sounds right to me.
